examples/diffusion/diffusion-cli.cpp

```diff
@@ -5,65 +5,192 @@
 #include "log.h"
 
 #include <limits.h>
-#include <string>
-#include <vector>
+
 #include <algorithm>
 #include <cmath>
+#include <cstring>
 #include <limits>
 #include <random>
+#include <string>
+#include <vector>
 
-typedef bool (*diffusion_step_callback_t)(int32_t step,
-                                          int32_t total_steps,
-                                          const llama_token * tokens,
-                                          int32_t n_tokens,
-                                          void * user_data);
-
-enum diffusion_alg {
-    DIFFUSION_ALG_ORIGIN       = 0,
-    DIFFUSION_ALG_MASKGIT_PLUS = 1,
-    DIFFUSION_ALG_TOPK_MARGIN  = 2,
-    DIFFUSION_ALG_ENTROPY      = 3,
+// Dream remasking algorithms
+enum diffusion_algorithm_dream {
+    ORIGIN       = 0,
+    MASKGIT_PLUS = 1,
+    TOPK_MARGIN  = 2,
+    ENTROPY      = 3,
+};
+
+// LLaDA remasking types
+enum diffusion_algorithm_llada {
+    LOW_CONFIDENCE = 0,
+    RANDOM         = 1,
 };
 
+typedef bool (*diffusion_step_callback_t)(int32_t             step,
+                                          int32_t             total_steps,
+                                          const llama_token * tokens,
+                                          int32_t             n_tokens,
+                                          void *              user_data);
+
+// Unified diffusion parameters structure
 struct diffusion_params {
+    diffusion_params() {
+        steps                   = 128;
+        temperature             = 0.2f;
+        mask_token_id           = LLAMA_TOKEN_NULL;
+        step_callback           = nullptr;
+        step_callback_user_data = nullptr;
+        seed                    = 0;
+    }
+
     int32_t                   steps;
-    float                     eps;
     float                     temperature;
-    float                     top_p;
-    int32_t                   top_k;
     llama_token               mask_token_id;
-    enum diffusion_alg        algorithm;
-    float                     alg_temp;
     diffusion_step_callback_t step_callback;
     void *                    step_callback_user_data;
     int32_t                   seed;
+    bool                      visual_mode;
+};
+
+struct dream_diffusion_params : diffusion_params {
+    float                          eps;
+    float                          top_p;
+    int32_t                        top_k;
+    enum diffusion_algorithm_dream algorithm;
+    float                          alg_temp;
 };
 
+struct llada_diffusion_params : diffusion_params {
+    int32_t                        max_length;
+    int32_t                        block_length;
```

Comment:
I don't think this separation of the diffusion parameters per architecture is necessary. It should be a single flat struct `diffusion_params` for all models.